README.md
Tomate Notify Plugin
====================

Tomate is a timer for the Pomodoro Technique.
This plugin shows OSD notifications when the timer starts and stops.

Installation
------------

Go to [this page](https://github.com/eliostvs/tomate-gtk) for instructions.

Bugs and Suggestions
-------------------

Bugs and suggestions should be reported [here](https://github.com/eliostvs/tomate-notify-plugin/issues).

License
-------

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License version 3, as published
by the Free Software Foundation.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranties of
MERCHANTABILITY, SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR
PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program.  If not, see <http://www.gnu.org/licenses/>.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.14.1

### Fixed

- Workaround setuptools autodiscovery modules

## 0.14.0

### Changed

- Tomate event API

## 0.13.0

### Fixes

- New tomate API

## 0.12.0

### Added

- Show notification when session stops

## 0.11.0

### Changed

- Remove infinity timeout of the notification

## 0.10.0

### Changed

- The timer, session and settings now emit a payload object instead of a dictionary

## 0.9.0

### Changed

- The Task enum was renamed to Sessions

## 0.8.0

- Python 3 only

## 0.7.0

- Not show notification when timer stopped

## 0.6.0

- Using py.test

## 0.5.0

- Remove linux package metadata
- Fix Gtk warnings

## 0.4.0

- Using the new event system


test.py
from __future__ import unicode_literals

import pytest
from mock import Mock, patch

from tomate.constant import Task
from tomate.graph import graph


@pytest.fixture()
def plugin():
    graph.register_factory('tomate.config', Mock)

    from notify_plugin import NotifyPlugin

    return NotifyPlugin()


@patch('gi.repository.Notify.init')
def test_init_dbus(init, plugin):
    plugin.activate()

    init.assert_called_with('Tomate')


@patch('gi.repository.Notify.uninit')
def test_uninit_dbus(uninit, plugin):
    plugin.deactivate()

    uninit.assert_called_with()


def test_get_icon_path(plugin):
    plugin.config.get_icon_path.return_value = '/path/to/mock/32/tomate.png'

    assert plugin.iconpath == '/path/to/mock/32/tomate.png'


@patch('gi.repository.Notify.Notification.new')
def test_should_show_pomodoro_start_session_message(notification, plugin):
    plugin.on_session_started()

    title = plugin.messages['pomodoro']['title']
    message = plugin.messages['pomodoro']['content']

    notification.assert_called_once_with(title, message, plugin.iconpath)


@patch('gi.repository.Notify.Notification.new')
def test_should_show_end_session_message(notification, plugin):
    plugin.on_session_ended(task=Task.shortbreak)

    notification.assert_called_once_with("The time is up!", '', plugin.iconpath)


tests.py
from __future__ import unicode_literals

import pytest
from mock import Mock, patch

from tomate.constant import Task, State
from tomate.graph import graph
from tomate.event import Events


def method_called(result):
    return result[0][0]


def setup_function(function):
    graph.providers.clear()

    graph.register_factory('tomate.config', Mock)

    Events['Session'].receivers.clear()


@pytest.fixture()
def plugin():

    from notify_plugin import NotifyPlugin

    return NotifyPlugin()


@patch('gi.repository.Notify.init')
def test_init_dbus(init, plugin):
    plugin.activate()

    init.assert_called_with('Tomate')


@patch('gi.repository.Notify.uninit')
def test_uninit_dbus(uninit, plugin):
    plugin.deactivate()

    uninit.assert_called_with()


def test_get_icon_path(plugin):
    plugin.config.get_icon_path.return_value = '/path/to/mock/32/tomate.png'

    assert plugin.iconpath == '/path/to/mock/32/tomate.png'


@patch('gi.repository.Notify.Notification.new')
def test_should_show_pomodoro_start_session_message(notification, plugin):
    plugin.on_session_started()

    title = plugin.messages['pomodoro']['title']
    message = plugin.messages['pomodoro']['content']

    notification.assert_called_once_with(title, message, plugin.iconpath)


@patch('gi.repository.Notify.Notification.new')
def test_should_show_session_finished_message(notification, plugin):
    plugin.on_session_finished(task=Task.shortbreak)

    notification.assert_called_once_with("The time is up!", '', plugin.iconpath)


def test_should_call_on_session_finished_when_session_finished(plugin):
    plugin.activate()

    result = Events['Session'].send(State.finished)

    assert len(result) == 1
    assert plugin.on_session_finished == method_called(result)


def test_should_call_on_session_started_when_session_started(plugin):
    plugin.activate()

    result = Events['Session'].send(State.started)

    assert len(result) == 1
    assert plugin.on_session_started == method_called(result)


test_plugin.py
from unittest.mock import Mock, patch

import gi
import pytest

from tomate.pomodoro import Sessions, State
from tomate.pomodoro.config import Config
from tomate.pomodoro.event import Events
from tomate.pomodoro.graph import graph
from tomate.pomodoro.session import Payload as SessionPayload

gi.require_version("Notify", "0.7")


def setup_function():
    graph.providers.clear()

    graph.register_instance(
        "tomate.config",
        Mock(spec=Config, **{"icon_path.return_value": "/path/to/mock/32/tomate.png"}),
    )

    Events.Session.receivers.clear()


@pytest.fixture()
@patch("gi.repository.Notify.Notification.new")
def plugin(_):
    from notify_plugin import NotifyPlugin

    return NotifyPlugin()


@patch("gi.repository.Notify.init")
def test_enable_notify_when_plugin_active(init, plugin):
    plugin.activate()

    init.assert_called_with("tomate-notify-plugin")


@patch("gi.repository.Notify.uninit")
def test_disable_notify_when_plugin_deactivate(uninit, plugin):
    plugin.deactivate()

    uninit.assert_called_with()


def test_show_notification_when_session_starts(plugin):
    plugin.activate()

    session_type = Sessions.pomodoro
    title = plugin.messages[session_type.name]["title"]
    message = plugin.messages[session_type.name]["content"]

    payload = SessionPayload(
        id="1234",
        state=State.started,
        type=Sessions.pomodoro,
        pomodoros=0,
        duration=25 * 60,
        task="",
    )

    Events.Session.send(State.started, payload=payload)

    plugin.notification.update.assert_called_once_with(
        title, message, "/path/to/mock/32/tomate.png"
    )

    plugin.notification.show.assert_called_once()


def test_show_notification_when_sessions_ends(plugin):
    plugin.activate()

    Events.Session.send(State.finished)

    plugin.notification.update.assert_called_once_with(
        "The time is up!", "", "/path/to/mock/32/tomate.png"
    )
    plugin.notification.show.assert_called_once()


def test_show_notification_when_session_stops(plugin):
    plugin.activate()

    Events.Session.send(State.stopped)

    plugin.notification.update.assert_called_once_with(
        "Session stopped manually", "", "/path/to/mock/32/tomate.png"
    )
    plugin.notification.show.assert_called_once()


pavement.py
#!/bin/env python
import os
from optparse import Option

from paver.easy import cmdopts, needs, path, sh
from paver.tasks import task

ROOT_PATH = path(__file__).dirname().abspath()

PLUGIN_PATH = ROOT_PATH / 'data' / 'plugins'

TOMATE_PATH = ROOT_PATH / 'tomate'


@task
@needs(['test'])
def default():
    pass


@task
@needs(['clean'])
@cmdopts([
    Option('-v', '--verbosity', default=1, type=int),
])
def test(options):
    os.environ['PYTHONPATH'] = '%s:%s' % (TOMATE_PATH, PLUGIN_PATH)
    sh('nosetests --cover-erase --with-coverage --verbosity=%s tests.py' % options.test.verbosity)


@task
def clean():
    sh('pyclean data/plugin')
    sh('pyclean .')
    sh('rm .coverage', ignore_error=True)


@task
@needs(['docker_rmi', 'docker_build', 'docker_run'])
def docker_test():
    pass


@task
def docker_rmi():
    sh('docker rmi eliostvs/tomate-notify-plugin', ignore_error=True)


@task
def docker_build():
    sh('docker build -t eliostvs/tomate-notify-plugin .')


@task
def docker_run():
    sh('docker run --rm eliostvs/tomate-notify-plugin')


data/plugins/notify_plugin.py
import logging
from locale import gettext as _
from typing import Tuple

import gi

gi.require_version("Notify", "0.7")

from wiring import Graph
from gi.repository import Notify

import tomate.pomodoro.plugin as plugin
from tomate.pomodoro import Bus, suppress_errors, Events, on, SessionType, SessionPayload

logger = logging.getLogger(__name__)


class NotifyPlugin(plugin.Plugin):
    messages = {
        SessionType.POMODORO: {"title": _("Pomodoro"), "content": _("Get back to work!")},
        SessionType.SHORT_BREAK: {"title": _("Short Break"), "content": _("It's coffee time!")},
        SessionType.LONG_BREAK: {"title": _("Long Break"), "content": _("Step away from the machine!")},
    }

    @suppress_errors
    def __init__(self):
        super().__init__()
        self.config = None
        self.notification = Notify.Notification.new("tomate-notify-plugin")

    def configure(self, bus: Bus, graph: Graph) -> None:
        super().configure(bus, graph)
        self.config = graph.get("tomate.config")

    @suppress_errors
    def activate(self):
        super().activate()
        Notify.init("tomate-notify-plugin")

    @suppress_errors
    def deactivate(self):
        super().deactivate()
        Notify.uninit()

    @on(Events.SESSION_START)
    def on_session_started(self, payload: SessionPayload):
        self.show_notification(*self.get_message(payload.type))

    @on(Events.SESSION_END)
    def on_session_finished(self,  **__):
        self.show_notification(title="The time is up!")

    @on(Events.SESSION_INTERRUPT)
    def on_session_stopped(self, **__):
        self.show_notification(title="Session stopped manually")

    def get_message(self, session: SessionType) -> Tuple[str, str]:
        return (
            self.messages[session]["title"],
            self.messages[session]["content"],
        )

    @suppress_errors
    def show_notification(self, title, message=""):
        self.notification.update(title, message, self.icon_path)
        result = self.notification.show()

        logger.debug(
            'action=show title="%s" message="%s" success=%r icon=%s',
            title,
            message,
            result,
            self.icon_path,
        )

    @property
    def icon_path(self):
        return self.config.icon_path("tomate", 32)
